### ISSUE TITLE
[Multiple Datasource] With default data source rendered in the data source component, we need to let consumers know which default data source id is used

### ISSUE THREAD

**Issue by BionIT** [opensearch-project/OpenSearch-Dashboards#6276]
**Is your feature request related to a problem? Please describe.**

In https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6237 and https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6186, we introduced new feature to setting default data source as well as to make sure there is a default data source. The default data source will be the default option in context where data source is needed. This introduces some complexity in the state management since only the data source component knows about the default data source and renders it as default option, we need to pass the default data source id back to the consumers

**Describe the solution you'd like**

We can use the callback function passed by the consumers at the first render when the default data source is identified.

**Describe alternatives you've considered**

Show empty page until user changes the data source selector

**Additional context**

Tasks:
- [x] data source selector should return the chosen default data source and dev tool and add sample data should use it to set the default data source: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6293
- [x] data source selectable should return the chosen default data source: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6327

**Comment by RamakrishnaChilaka** [comment]
@BionIT, letting the consumer know via callback approach will result in two renders, and the backend APIs will be called from the first render already which we want to avoid, as it is unnecessary traffic.

If it is possible, we can expose a function in the core, which says what is the default data source id or the default data source is, we can initialise the default state in our constructor to that. If the CX gave no default data source, you initialise it to Local Cluster. 

I am not sure how can we handle if `hideLocalCluster` prop as true, but the dataSourceId passed to the component is empty..

**Comment by BionIT** [comment]
@RamakrishnaChilaka  Based on discussion, the plugin page which consumes the component, should render the data source picker first, and the picker will have the logic to determine appropriate default option and use the callback function to inform the page, the rest of page should wait until the initial value is retrieved before rendering, once the value is received, then render the rest of page

The default/initial option provided back by the picker would have 3 possible states with default data source introduced and no local cluster scenario:
1. undefined - this means there is no default data source, no local cluster, and the page should not make any external requests, and the page should render no results (we need to check how it would look like)
2. empty data source id {id:'', label: 'Local cluster'} - this means there is no default data source, but has local cluster, and the page should make request to local cluster
3. data source id {id:'<data source id>', label: '<data source label>'} - this means there is a data source being recognized as default by the picker logic, and the page should make request to the remote data source using the data source id

**Comment by RamakrishnaChilaka** [comment]
Thanks Lu. @BionIT, can you provide a utility function that takes a `dataSourceId` as input and returns whether the plugin code can be mounted? Because it's better to not have context inside Plugin code that when `dataSourceId` is undefined, plugin will mount its body and we will repeat this logic across all plugins, which is not desirable.

or provide with an additional boolean field in `dataSourcecallBackFunc` like `canMountPluginComponent`, when it's true we will mount the rest of the plugin code, otherwise we will render Opensearch loader or some other component based on the UI/UX feedback.

**Comment by kavilla** [comment]
Without diving into the original implementation. I have a few questions.

@BionIT, does this work in the same way that default index patterns work? The initial data source will be the first data source? Does it get saved into the config (UI Settings) saved object? For example like:

![Screenshot 2024-04-01 at 6 56 22 PM](https://github.com/opensearch-project/OpenSearch-Dashboards/assets/12060704/72649b62-163c-4528-a569-15ea035b3282)

If not then I think we should follow the existing method of adding the UI setting for this? This setting can be added if data sources is enabled, and if the default data source is null then maybe that can be the indication that the user has data sources enabled and we render it.

Also we can consider adding the default data source to add to the app state `_a` so on load we do a request to get the default data source if enabled and then populate the app state with it and if the plugin needs checks the app state and sees a defined data source id then it will get render that logic so.

Basically mimicking the index pattern, here we can see the app state that includes the index pattern id: 
```
https://playground.opensearch.org/app/data-explorer/discover#?_a=(discover:(columns:!(_source),isDirty:!f,sort:!()),metadata:(indexPattern:ff959d40-b880-11e8-a6d9-e546fe2bba5f,...
```

Please ignore if this does not actually relate to the issue you are trying to solve.

**Comment by kavilla** [comment]
Will assign you @BionIT since you open the issue as it seems like a RFC. If this is incorrect please unassign but will remove the Untriage label.

**Comment by BionIT** [comment]
Thanks @kavilla ! @zhyuanqi is working on it, and default data source is implemented similarly as index pattern inside uiSetting, see implementations in https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6237 and https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6186 . The complexity is default data source might not be the default option shown in the data source component since it might be an incompatible data source which based on requirement should not be shown. Thus the default data source set by user might not be the default option shown to the user, and we need to let plugin know since plugin pages make requests to backend to fetch results based on selected data source on initial page load - cc @kgcreative

**Comment by zhyuanqi** [comment]
working on it.

**Comment by BionIT** [comment]
Thanks @zhyuanqi !!!

**Comment by BionIT** [comment]
2 PRs have been merged, closing the issue
